src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx

```diff
@@ -0,0 +1,953 @@
+          return data;
+        }
+        return rel;
+      });
+    }
+    setDatasetSchema(schemaCopy);
+    setRelationshipModalOpen(false);
+  };
+
+  const deleteRelationship = () => {
+    const schemaCopy = _.cloneDeep(datasetSchema);
+    schemaCopy.relationships = schemaCopy.relationships?.filter(
+      (rel: RelationshipModel) => rel.name !== relationshipModalDefaultValues.name,
+    );
+    setDatasetSchema(schemaCopy);
+    setRelationshipModalOpen(false);
+  };
+
+  const highlightRelationshipColumns = (rel: RelationshipModel) => {
+    const fromIndices = getIndices(rel.from.table, rel.from.column);
+    const toIndices = getIndices(rel.to.table, rel.to.column);
+    setOutlinedRelationships([
+      `${fromIndices.table}-${fromIndices.column}`,
+      `${toIndices.table}-${toIndices.column}`,
+    ]);
+  };
+
+  // ----------------------------------------
+  // Table Helper Methods
+  // ----------------------------------------
+  const getIndices = (table: string, column: string) => {
+    const tableIndex = _.findIndex(
+      datasetSchema.tables,
+      (schemaTable: TableModel) => schemaTable.name === table,
+    );
+    const columnIndex = _.findIndex(
+      datasetSchema.tables[tableIndex].columns,
+      (schemaCol: ColumnModel) => schemaCol.name === column,
+    );
+    return {
+      table: tableIndex,
+      column: columnIndex,
+    };
+  };
+
+  const getNodeId = (tableIndex: number, columnIndex: number) => {
+    if (tableIndex === -1) {
+      return '';
+    }
+    if (columnIndex > -1) {
+      return `${tableIndex}-${columnIndex}`;
+    }
+    return `${tableIndex}`;
+  };
+
+  const swapArrayLocs = (arr: Array<any>, index1: number, index2: number) => {
+    [arr[index1], arr[index2]] = [arr[index2], arr[index1]];
+  };
+
+  const moveSelectedUp = () => {
+    const schemaCopy = _.cloneDeep(datasetSchema);
+    if (selectedColumn > 0) {
+      swapArrayLocs(schemaCopy.tables[selectedTable].columns, selectedColumn, selectedColumn - 1);
+      setSelectedColumn(selectedColumn - 1);
+    } else if (selectedColumn === -1 && selectedTable > 0) {
+      swapArrayLocs(schemaCopy.tables, selectedTable, selectedTable - 1);
+      setSelectedTable(selectedTable - 1);
+    }
+    setDatasetSchema(schemaCopy);
+  };
+
+  const moveSelectedDown = () => {
+    const schemaCopy = _.cloneDeep(datasetSchema);
+    if (
+      selectedColumn !== -1 &&
+      selectedColumn < schemaCopy.tables[selectedTable].columns.length - 1
+    ) {
+      swapArrayLocs(schemaCopy.tables[selectedTable].columns, selectedColumn, selectedColumn + 1);
+      setSelectedColumn(selectedColumn + 1);
+    } else if (selectedTable !== -1 && selectedTable < schemaCopy.tables.length - 1) {
+      swapArrayLocs(schemaCopy.tables, selectedTable, selectedTable + 1);
+      setSelectedTable(selectedTable + 1);
+    }
+    setDatasetSchema(schemaCopy);
+  };
+
+  // ----------------------------------------
+  // CodeMirror (JSON viewer)
```

Comment:
The CodeMirror is super cool.  I wonder if we want to "lock" it on default, though?  I could see a user wanting to create the schema via the GUI, inadvertently modifying the mirrored JSON, and not being clear on how to fix.  (For example, you could change the JSON for a column definition to illegally stack several types: it would then reflect in the GUI but you'd be unable to fix it there.)

<img width="1477" alt="Screen Shot 2023-01-20 at 12 22 22 PM" src="https://user-images.githubusercontent.com/79769153/213799975-2be77bd2-9945-4f00-9ddf-d58b5d7155db.png">